librepo/handle.c

```diff
@@ -264,6 +266,32 @@ lr_handle_remote_sources_changed(LrHandle *handle, LrChangedRemoteSource type)
     }
 }
 
+struct callback_data {
+   GMainLoop *loop; 
+   long input_seconds;
+   time_t begin;
+   GNetworkMonitor *monitor;
+   GSocketConnectable *connectable; 
+   GCancellable *cancellable;
+};
+
+gboolean timeout_callback(gpointer data)
+{ 
+  struct callback_data *dt = (struct callback_data*)data;
+   if(!g_network_monitor_get_network_available(dt->monitor) &&
+      g_get_monotonic_time() - dt->begin <= dt->input_seconds * 1000000){
```

Comment:
~1000000 here is `G_USEC_PER_SEC`; it may be simpler actually to do `g_get_monotonic_time() / G_USEC_PER_SEC` instead.~

EDIT: Per above, then we can just do `if (g_get_monotonic_time () > dt->deadline)`.